manpages/fproxyproxy.1
.TH "fproxyproxy" "1" "0.2.2" "Freenet Project" "pyFreenet"
.SH "NAME"
.LP 
fproxyproxy \- a name\-service\-enabled http proxy service over fproxy
.SH "SYNOPSIS"
.LP 
\fBfproxyproxy\fP [\fIoptions\fP]
.SH "DESCRIPTION"
.LP 
\fBfproxyproxy\fR is an http proxy server that sits on top of
Freenet's FProxy server. It translates human\-friendly
freenet 'domain names' received from the client into proper
freenet URIs, feeds these URIs into fproxy, and feeds FProxy's
output back to the client.
.LP
Whereas FProxy is designed to be used as a web server,
fproxyproxy is designed to be configured into the browser as
an http proxy server. Fortunately, with excellent quick
proxy\-switching features of modern browsers (such as the
\fBSwitchProxy\fR extension in \fBFirefox\fR), this
becomes an easy and simple operation.
.LP
As an example of the use of fproxyproxy,
if you had a 'freenet name services' record
for the 'pyFreenet' website, you would be able to type
\fBhttp://pyFreenet\fR
into your browser, and be viewing the pyFreenet freesite! (As long
as you have set up fproxyproxy as an http proxy in your web browser).
.LP
fproxyproxy is a proof of concept for using secure, human\-friendly
freesite addresses within freenet, and is intended to be used
alongside the \fBfcpnames(1)\fR utility.
.LP
If you want to access FProxy directly in a browser configured to use
fproxyproxy as a proxy server, just enter the address
\fBhttp://fproxy\fR.
.SH "OPTIONS"
.TP 
\fB\-h\fR, \fB\-\-help\fR
Print help information and exit
.TP 
\fB\-V\fR, \fB\-\-version\fR
Print version number and exit
.TP 
\fB\-v\fR, \fB\-\-verbose\fR
Enable very verbose output, including FCP transcripts,
to the stderr stream.
.TP 
\fB\-H\fR, \fB\-\-fcpHost=\fIhostname\fR
Use FCP interface at host \fIhostname\fR,
defaults to \fB127.0.0.1\fR
.TP 
\fB\-P\fR, \fB\-\-fcpPort=\fIport\fR
Use FCP interface at port \fIport\fR
.TP 
\fB\-p\fR, \fB\-\-fproxyAddress=\fI[hostname][:port]\fR
Point fproxyproxy at the FProxy service at address
\fIhostname:port\fR, default is \fB127.0.0.1:8888\fR.
.TP 
\fB\-L\fR, \fB\-\-listenAddress=\fI[hostname][:port]\fR
Sets the address fproxyproxy listens on for client http
connections, default is \fB127.0.0.1:8889\fR.
.SH "ENVIRONMENT VARIABLES"
.TP 
\fBFCP_HOST\fP
Specifies the hostname of the FCP interface. This value
will be used unless overridden with '\-H'.
.TP 
\fBFCP_PORT\fP
Specifies the port number of the FCP interface. This value
will be used unless overridden with '\-P'.
.SH "EXAMPLES"
.TP 
\fBfproxyproxy \-L localhost:2225 \-p 192.168.1.5\fP
Run fproxyproxy, listening for client connections on
\fBlocalhost:2225\fR, and forwarding requests to
an FProxy port at \fB192.168.1.5:8888\fR (note the
default 8888 being applied to fproxy port).
.SH "BUGS"
.LP
This man page documents an older version of fproxyproxy. Some
informations may be missing or wrong.
.SH "REPORTING BUGS"
.LP
If you find a bug, please post it to the Freenet bug tracker at:
.RS
.LP    
https://bugs.freenetproject.org/bug_report_page.php
.RE
.LP
but first please be sure to target the 'pyFreenet' project, by
selecting 'pyFreenet' in the combobox on the top-right of the page.
.SH "AUTHORS"
.LP
The pyFreenet applications and library were originally written
by David McNab (aum) in May-June, 2006 and are maintained
by the Freenet Project team.
.LP
Contact the original author at: <davidmcnabnz@gmail.com>
.SH "SEE ALSO"
.LP 
fcpnames(1)


CHANGELOG

Revision history for pyFreenet

- Version 0.2.5
    - added 'noderefbot' - an IRC bot which swaps noderefs automatically

- Version 0.2.4

    - minor bugfix release
    - freesitemgr: now saving updated freesite edition numbers in
      ~/.freesitemgr/<sitename>

- Version 0.2.3

    - added 'fproxyproxy' shim atop fproxy, which should be used as
      a browser proxy, not surfed directly

- Version 0.2.2

    - added 'name services layer' (see 'fcpnames' manpage)
    - integrated this layer into .get()/.put() methods (and
      consequently, into fcpget and fcpput
    - added the utility 'fcpnames', allowing easy management
      of name services

- Version 0.2.1

    - added commands 'help' and 'cleanup' to freesitemgr
    - moved fcpget/fcpput/etc into modules in package fcp, leaving simple
      2-liner front-ends

- Version 0.2

    - added another layer of recovery to freesitemgr, to re-submit
      jobs the node forgets about (ie, losts from its queue)
    - added '-C' option to freesitemgr, to suit its use in crontab entries

- Version 0.1.8

    - another rework of freesitemgr:
        - inserts use only the global queue
        - 'freesitemgr update' exits immediately, leaving jobs on queue
        - extensive accounting, which recognises and compensates for node
          failures
        - you can now safely cron 'freesitemgr update'

- Version 0.1.7

    - rewrote most of freesitemgr:
        - now uses directory ~/.freesitemgr/
        - keeps track of all inserts
        - if interrupted, or if node dies, will resume where
          it left off
        - simplified options

- Version 0.1.6

    - now storing manifests in .freesites (for freesitemgr)
    - freesitemgr now issues progress messages with '-v' set
    - all utils now accept multiple occurrences of '-v' option, with
      increasing verbosity
    - added timeout option to get/put (and '-t' to fcpget/fcpput)

- Version 0.1.5

    - added global queue and persistence support for fcpget/fcpput
    - added 'nowait' option to fcpput

- Version 0.1.4

    - added manpages for console programs
    - improved mimetype determination algorithm for fcp put
    - added 'freedisk', a rudimentary linux filesystem that maps freenet
      into a mountable fs - limited functionality at present
    - added support for use of env vars FCP_HOST and FCP_PORT for specifying
      FCP host/port - useful for people who access FCP across a LAN, since
      it avoids annoyance of having to specify -H or -P with each command

- Version 0.1.3

    - added 'fcpget' and 'fcpput' command-line key retrieve/insert apps
    - added 'freesitemgr' command-line freesite insertion app
    - several bug fixes

- Version 0.1.2

    - added xmlrpc server app
    - added xmlrpc server CGI module (for embedding a Freenet XML-RPC
      server into websites)
    - added 'freesitemgr', a console freesite insertion app

- Version 0.1.1
    - 2006-May-13
        - First packaged release



fcp/fproxyproxy.py
#@+leo-ver=4
#@+node:@file fproxyproxy.py
"""
fproxyproxy

An http proxy atop fproxy which uses pyFreenet's 'name services'
"""

#@+others
#@+node:imports
import sys, os, getopt, traceback, mimetypes, time
from BaseHTTPServer import HTTPServer
from SimpleHTTPServer import SimpleHTTPRequestHandler
from SocketServer import ThreadingMixIn
from httplib import HTTPConnection
import socket

import node
from node import ERROR, INFO, DETAIL, DEBUG

#@-node:imports
#@+node:globals
progname = sys.argv[0]

#@-node:globals
#@+node:class Handler
class Handler(SimpleHTTPRequestHandler):
    """
    Handles each FProxyProxy client request
    """
    #@    @+others
    #@+node:__init__
    def __init__(self, request, client_address, server):
        
        self.server = server
        SimpleHTTPRequestHandler.__init__(self, request, client_address, server)
    
    #@-node:__init__
    #@+node:do_GET
    def do_GET(self):
        
        print "GET: client=%s path=%s" % (self.client_address, self.path)
    
        #SimpleHTTPRequestHandler.do_GET(self)
    
        try:
            #mimetype, data = self.server.node.get(self.path[1:])
            #self.send_response(200)
            #self.send_header("Content-type", mimetype)
            #self.send_header("Content-Length", str(len(data)))
            #self.end_headers()
            #self.wfile.write(data)
            #self.wfile.flush()
    
            self.fproxyGet(self.path)
    
        except:
            traceback.print_exc()
            self.send_error(404, "File not found")
            return None
    
    #@-node:do_GET
    #@+node:fproxyGet
    def fproxyGet(self, path):
        """
        Fetches from fproxy, returns (status, mimetype, data)
        """
        server = self.server
        headers = self.headers
    
        print "--------------------------------------------"
        print "** path=%s" % path
    
        # first scenario - user is pointing their browser directly at
        # fproxyfproxy, barf!
        if not path.startswith("http://"):
            self.send_response(400)
            self.send_header("Content-type", "text/html")
            data = "\n".join([
                "<html><head>",
                "<title>Access Denied</title>",
                "</head><body>",
                "<h1>Access Denied</h1>",
                "Sorry, but FProxyProxy is an http proxy server.<br>",
                "Please don't try to access it like a web server.",
                "</body></html>",
                "",
                ])
            self.send_header("Content-Length", str(len(data)))
            self.send_header("Location", location)
            self.end_headers()
            self.wfile.write(data)
            self.wfile.flush()
            return
    
        # convert path to relative
        path = "/" + path[7:].split("/", 1)[-1]
        #print "** path=%s" % repr(path)
    
    
        try:
            # check host header
            hostname = headers.get("Host", 'fproxy')
            pathbits = path.split("/")
    
            print "** hostname = %s" % hostname
    
            # second scenario, user has just given a domain name without trailing /
            if len(pathbits) == 1:
                # redirect to force trailing slash
                location = path + "/"
                print "** redirecting to: %s" % location
    
                self.send_response(301)
                self.send_header("Content-type", "text/html")
                data = "\n".join([
                    "<html><head>",
                    "<title>Permanent redirect: new URI</title>",
                    "</head><body>",
                    "<h1>Permanent redirect: new URI</h1>",
                    "<a href=\"%s\">Click here</a>",
                    "</body></html>",
                    "",
                    ]) % location
                self.send_header("Content-Length", str(len(data)))
                self.send_header("Location", location)
                self.end_headers()
                self.wfile.write(data)
                self.wfile.flush()
                return
    
            tail = "/".join(pathbits[1:])
    
            # third scenario - request into fproxy
            if hostname == 'fproxy':
    
                # tis an fproxy request, go straight through
                conn = HTTPConnection(server.fproxyHost, server.fproxyPort)
                conn.request("GET", path)
                resp = conn.getresponse()
                self.send_response(resp.status)
                self.send_header("Content-type",
                                 resp.getheader("Content-Type", "text/plain"))
                data = resp.read()
                self.send_header("Content-Length", str(len(data)))
                self.end_headers()
                self.wfile.write(data)
                self.wfile.flush()
                conn.close()
                return
    
            else:
                # final scenario - some other domain, try lookup
                uri = server.node.namesiteLookup(hostname)
    
                if not uri:
                    # lookup failed, do the usual 404 thang
                    print "** lookup of domain %s failed" % hostname
                    self.send_response(404)
                    self.send_header("Content-type", "text/html")
                    data = "\n".join([
                        "<html><head>",
                        "<title>404 - Freenet name not found</title>",
                        "</head><body",
                        "<h1>404 - Freenet name not found</title>",
                        "The pyFreenet name service was unable to resolve ",
                        "the name %s" % hostname,
                        "<br><br>",
                        "You might like to find its freenet uri and try that ",
                        "within <a href=\"/fproxy/\">FProxy</a>",
                        "</body></html>",
                        "",
                        ])
                    self.send_header("Content-Length", str(len(data)))
                    self.end_headers()
                    self.wfile.write(data)
                    self.wfile.flush()
    
                # lookup succeeded - ok to go now via fproxy
                conn = HTTPConnection(server.fproxyHost, server.fproxyPort)
                newpath = "/" + uri
                if tail:
                    if not newpath.endswith("/"):
                        newpath += "/"
                    newpath += tail
                print "** newpath=%s" % newpath
                conn.request("GET", newpath)
                resp = conn.getresponse()
                print "** status=%s" % resp.status
                self.send_response(resp.status)
                self.send_header("Content-type",
                                 resp.getheader("Content-Type", "text/plain"))
    
                # has fproxy sent us a redirect?
                if resp.status == 301:
                    # yuck, fproxy is telling us to redirect, which
                    # sadly means we have to lose the domain name
                    # from our browser address bar
                    location = resp.getheader("location")
                    newLocation = "http://fproxy" + location
                    print "*** redirected!!!"
                    print "*** old location = %s" % location
                    print "***  --> %s" % newLocation
                    self.send_header("Location", newLocation)
    
                # get the data from fproxy and send it up to the client
                data = resp.read()
                self.send_header("Content-Length", str(len(data)))
                self.end_headers()
                self.wfile.write(data)
                self.wfile.flush()
                conn.close()
                return
    
            return
    
        except socket.error:
            raise
    
    #@-node:fproxyGet
    #@-others

#@-node:class Handler
#@+node:class FProxyProxy
class FProxyProxy(ThreadingMixIn, HTTPServer):
    """
    an http proxy that runs atop fproxy, and uses the pyFreenet name service
    """
    #@    @+others
    #@+node:__init__
    def __init__(self, **kw):
        """
        runs the FProxyProxy service
        
        Keywords:
            - node - a live FCPNode object
            - fproxyHost - hostname of fproxy
            - fproxyPort - port of fproxy
            - listenHost - hostname to listen on for client HTTP connections
            - listenPort - port to listen on for client HTTP connections
        """
        for k in ['node', 'fproxyHost', 'fproxyPort', 'listenHost', 'listenPort']:
            setattr(self, k, kw[k])
    
        self.log = self.node._log
    
        HTTPServer.__init__(self, (self.listenHost, self.listenPort), Handler)
    
    #@-node:__init__
    #@+node:run
    def run(self):
        """
        Starts the proxy, runs forever till interrupted
        """
        log = self.log
        log(ERROR, "FproxyProxy listening on %s:%s" % (self.listenHost, self.listenPort))
        log(ERROR, "  -> forwarding requests to fproxy at %s:%s" % (
                    self.fproxyHost, self.fproxyPort))
    
        self.serve_forever()
    
    #@-node:run
    #@-others

#@-node:class FProxyProxy
#@+node:usage
def usage(msg=None, ret=1):
    """
    Prints usage message then exits
    """
    if msg:
        sys.stderr.write(msg+"\n")
    sys.stderr.write("Usage: %s [options] src-uri target-uri\n" % progname)
    sys.stderr.write("Type '%s -h' for help\n" % progname)
    sys.exit(ret)

#@-node:usage
#@+node:help
def help():
    """
    print help options, then exit
    """
    print "%s: runs an http proxy atop fproxy,"  % progname
    print "which uses pyFreenet 'name services'"
    print
    print "Note - you should configure fproxyproxy as an http proxy"
    print "in your browser (best done via Firefox's 'switchproxy' extension"
    print
    print "Usage: %s [options] src-uri target-uri" % progname
    print
    print "Options:"
    print "  -h, -?, --help"
    print "     Print this help message"
    print "  -v, --verbose"
    print "     Print verbose progress messages to stderr"
    print "  -H, --fcpHost=<hostname>"
    print "     Connect to FCP service at host <hostname>"
    print "  -P, --fcpPort=<portnum>"
    print "     Connect to FCP service at port <portnum>"
    print "  -p, --fproxyAddress=[<hostname>][:<portnum>]"
    print "     Use fproxy service at <hostname>:<portnum>,"
    print "     default 127.0.0.1:8888"
    print "  -L, --listenAddress=[<hostname>][:<portnum>]"
    print "     Listen for http connections on <hostname>:<portnum>,"
    print "     default is 127.0.0.1:8889"
    print "  -V, --version"
    print "     Print version number and exit"
    print
    print "Environment:"
    print "  Instead of specifying -H and/or -P, you can define the environment"
    print "  variables FCP_HOST and/or FCP_PORT respectively"

    sys.exit(0)

#@-node:help
#@+node:main
def main():
    """
    Front end for fproxyproxy utility
    """
    # default job options
    verbosity = node.ERROR
    verbose = False
    fcpHost = node.defaultFCPHost
    fcpPort = node.defaultFCPPort
    fproxyHost = os.environ.get("FPROXY_HOST", "127.0.0.1")
    fproxyPort = int(os.environ.get("FPROXY_PORT", 8888))
    listenHost = os.environ.get("FPROXYPROXY_HOST", "127.0.0.1")
    listenPort = int(os.environ.get("FPROXYPROXY_PORT", 8889))

    opts = {
            "Verbosity" : 0,
            }

    # process command line switches
    try:
        cmdopts, args = getopt.getopt(
            sys.argv[1:],
            "?hvH:P:Vp:L:",
            ["help", "verbose", "fcpHost=", "fcpPort=", "version",
             "listenAddress=", "fproxyAddress=",
             ]
            )
    except getopt.GetoptError:
        # print help information and exit:
        usage()
        sys.exit(2)
    output = None
    verbose = False
    #print cmdopts
    for o, a in cmdopts:

        if o in ("-?", "-h", "--help"):
            help()

        if o in ("-V", "--version"):
            print "This is %s, version %s" % (progname, node.fcpVersion)
            sys.exit(0)

        if o in ("-v", "--verbosity"):
            verbosity = node.DETAIL
            opts['Verbosity'] = 1023
            verbose = True

        if o in ("-H", "--fcpHost"):
            fcpHost = a
        
        if o in ("-P", "--fcpPort"):
            try:
                fcpPort = int(a)
            except:
                usage("Invalid fcpPort argument %s" % repr(a))

        if o in ("-L", "--listenAddress"):
            parts = a.split(":")
            if len(parts) == 1:
                listenHost = parts[0]
            elif len(parts) == 2:
                if parts[0]:
                    listenHost = parts[0]
                if parts[1]:
                    listenPort = int(parts[1])
            else:
                usage("Invalid listen address '%s'" % a)

        if o in ("-p", "--fproxyAddress"):
            parts = a.split(":")
            if len(parts) == 1:
                fproxyHost = parts[0]
            elif len(parts) == 2:
                if parts[0]:
                    fproxyHost = parts[0]
                if parts[1]:
                    fproxyPort = int(parts[1])
            
    # try to create an FCP node, needed for name lookups
    try:
        n = node.FCPNode(host=fcpHost, port=fcpPort, verbosity=verbosity,
                         logfile=sys.stderr)
        log = n._log
    except:
        if verbose:
            traceback.print_exc(file=sys.stderr)
        usage("Failed to connect to FCP service at %s:%s" % (fcpHost, fcpPort))

    try:
        proxy = FProxyProxy(
                    node=n,
                    fproxyHost=fproxyHost, fproxyPort=fproxyPort,
                    listenHost=listenHost, listenPort=listenPort)
        proxy.run()
        sys.exit(0)
    except KeyboardInterrupt:
        print "fproxyproxy terminated by user"
        n.shutdown()
        sys.exit(0)
    except:
        traceback.print_exc()
        print "fproxyproxy terminated"
        n.shutdown()
        sys.exit(1)

#@-node:main
#@-others

#@-node:@file fproxyproxy.py
#@-leo
